subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -51,44 +51,44 @@ public DependentsSet getRelevantDependents(Iterable<String> classes, IntSet cons
 
     public DependentsSet getRelevantDependents(String className, IntSet constants) {
         DependentsSet deps = data.getDependents(className);
-        if (deps != null && deps.isDependencyToAll()) {
+        if (deps.isDependencyToAll()) {
             return deps;
         }
-        if (deps == null && constants.isEmpty()) {
-            return DefaultDependentsSet.EMPTY;
-        }
         if (!constants.isEmpty()) {
             return DependencyToAll.INSTANCE;
         }
-        Set<String> result = new HashSet<String>();
-        if (deps != null && !deps.isDependencyToAll()) {
-            recurseDependents(new HashSet<String>(), result, deps.getDependentClasses());
+        if (deps.getDependentClasses().isEmpty()) {
+            return DefaultDependentsSet.EMPTY;
         }
+        Set<String> result = new HashSet<String>();
+        recurseDependents(new HashSet<String>(), result, deps.getDependentClasses());
         result.remove(className);
         return new DefaultDependentsSet(result);
-
     }
 
     public boolean isDependencyToAll(String className) {
-        DependentsSet deps = data.getDependents(className);
-        return deps != null && deps.isDependencyToAll();
+        return data.getDependents(className).isDependencyToAll();
     }
 
     private void recurseDependents(Set<String> visited, Set<String> result, Set<String> dependentClasses) {
         for (String d : dependentClasses) {
             if (!visited.add(d)) {
                 continue;
             }
-            if (!d.contains("$")) { //filter out the inner classes
+            if (!isInnerClass(d)) {
                 result.add(d);
             }
             DependentsSet currentDependents = data.getDependents(d);
-            if (currentDependents != null && !currentDependents.isDependencyToAll()) {
+            if (!currentDependents.isDependencyToAll()) {
                 recurseDependents(visited, result, currentDependents.getDependentClasses());
             }
         }
     }
 
+    private boolean isInnerClass(String d) {
```

Comment:
I know this is just refactored, but non-inner classes can contain dollar sign, right?

Reply:
That's right, this is just a heuristic, which we also use in the code which deletes stale class files. We could be more exact here, but it wouldn't make much of a difference in performance, so we just go a little more coarse grained.

I'll rename this method to make that clearer.